File: Sources/DLL/Arbre/Visiteurs/VisiteurObtenirSelection.cpp
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurObtenirSelection.cpp
/// @author Nam Lesage
/// @date 2016-09-27
/// @version 0.1
///
/// @addtogroup inf2990 INF2990
/// @{
///////////////////////////////////////////////////////////////////////////////

#include "VisiteurObtenirSelection.h"

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::VisiteurObtenirSelection()
///
/// Constructeur de VisiteurObtenirSelection
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurObtenirSelection::VisiteurObtenirSelection() {

}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::~VisiteurObtenirSelection()
///
/// Destructeur de VisiteurObtenirSelection
///
///	@param[in] Aucun
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
VisiteurObtenirSelection::~VisiteurObtenirSelection() {

}

////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::visiterAccelerateur(NoeudAccelerateur* noeud)
///
/// Cette fonction permet de visiteur un accélérateur pour la sélection. 
/// Si le noeud est sélectionné, on l'ajoute au vecteur.
///
///	@param[in] noeud : Un accélérateur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurObtenirSelection::visiterAccelerateur(NoeudAccelerateur* noeud) {
	if (noeud->estSelectionne()) {
		noeuds_.push_back(noeud);
	}
}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::visiterPointControl(NoeudPointControl * noeud)
///
/// Cette fonction permet de visiteur un point de contrôle pour la sélection. 
/// Si le noeud est sélectionné, on l'ajoute au vecteur.
///
///	@param[in] noeud : Un point de contrôle
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurObtenirSelection::visiterPointControl(NoeudPointControl * noeud)
{
	if (noeud->estSelectionne()) {
		noeuds_.push_back(noeud);
	}
}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::visiterMur(NoeudMur* noeud)
///
/// Cette fonction permet de visiteur un mur pour la sélection. 
/// Si le noeud est sélectionné, on l'ajoute au vecteur.
///
///	@param[in] noeud : Un mur
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurObtenirSelection::visiterMur(NoeudMur* noeud) {
	if (noeud->estSelectionne()) {
		noeuds_.push_back(noeud);
	}
}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::visiterPortail(NoeudPortail* noeud)
///
/// Cette fonction permet de visiteur un portail pour la sélection. 
/// Si le noeud est sélectionné, on l'ajoute au vecteur.
///
///	@param[in] noeud : Un portail
///
/// @return Aucune
///
////////////////////////////////////////////////////////////////////////
void VisiteurObtenirSelection::visiterPortail(NoeudPortail* noeud) {
	if (noeud->estSelectionne()) {
		noeuds_.push_back(noeud);
	}
}


////////////////////////////////////////////////////////////////////////
///
/// @fn VisiteurObtenirSelection::obtenirNoeuds()
///
/// Cette fonction retourne le vecteur de noeuds.
///
/// @return noeuds_
///
////////////////////////////////////////////////////////////////////////
std::vector<NoeudAbstrait*> VisiteurObtenirSelection::obtenirNoeuds() {
	return noeuds_;
}


////////////////////////////////////////////////////////////////////////
///
/// @fn glm::vec3 VisiteurObtenirSelection::obtenirCentreSelection()
///
/// Cette fonction calcule et retourne le centre de la sélection.
///
/// @return Le centre de la sélection
///
////////////////////////////////////////////////////////////////////////
glm::dvec3 VisiteurObtenirSelection::obtenirCentreSelection() {
	glm::dvec3 centre(0);

	for (auto noeud : noeuds_) {
		centre += noeud->obtenirPositionRelative();
	}

	if (noeuds_.size()) {
		centre /= noeuds_.size();
	}

	return centre;
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////

File: AirHockeyLightClient/AirHockeyLightClient/Noyau/Arbre/Visiteurs/VisiteurObtenirSelection.swift
///////////////////////////////////////////////////////////////////////////////
/// @file VisiteurObtenirSelection.swift
/// @author Pierre To
/// @date 2017-10-22
/// @version 1
///
/// @addtogroup log3900 LOG3990
/// @{
///////////////////////////////////////////////////////////////////////////////

import GLKit
import SceneKit

///////////////////////////////////////////////////////////////////////////
/// @class VisiteurObtenirSelection
/// @brief Permet d'obtenir tous les noeuds sélectionnés
///
/// @author Pierre To
/// @date 2017-10-22
///////////////////////////////////////////////////////////////////////////
class VisiteurObtenirSelection: VisiteurAbstrait {

    /// Vecteur regroupant les noeuds spécifiques
    private var noeuds = [NoeudCommun]()
    
    /// Constructeur
    init() {
    }
    
    /// Cette fonction retourne le vecteur de noeuds sélectionnés
    func obtenirNoeuds() -> [NoeudCommun] {
        return self.noeuds
    }
    
    /// Visiter un accéléateur pour l'obtention de la sélection
    func visiterAccelerateur(noeud: NoeudAccelerateur) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un maillet pour l'obtention de la sélection
    //virtual void visiterMaillet(NoeudMaillet* noeud);
    
    /// Visiter une table pour l'obtention de la sélection
    func visiterTable(noeud: NoeudTable) {
        // Ne fait rien
    }
    
    /// Visiter un point de contrôle pour l'obtention de la sélection
    func visiterPointControl(noeud: NoeudPointControl) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un mur pour l'obtention de la sélection
    func visiterMur(noeud: NoeudMur) {
        self.implementationDefaut(noeud: noeud)
    }
    
    /// Visiter un portail pour l'obtention de la sélection
    func visiterPortail(noeud: NoeudPortail) {
        self.implementationDefaut(noeud: noeud)
    }
    
    //virtual void visiterRondelle(NoeudRondelle* noeud);
    
    /// Si le noeud est sélectionné, on l'ajoute au vecteur.
    private func implementationDefaut(noeud: NoeudCommun) {
        if (noeud.estSelectionne()) {
            noeuds.append(noeud)
        }
    }
    
}

///////////////////////////////////////////////////////////////////////////////
/// @}
///////////////////////////////////////////////////////////////////////////////
